File: leetcode/0134-gas-station/solution.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        """
        RUNTIME: 654 ms (97.45%)
        MEMORY: 19.3 MB (23.25%)
        """
        total, tank, start = 0, 0, 0
        for i in range(len(gas)):
            tank = tank + gas[i] - cost[i]
            if tank < 0:
                start = i + 1
                total += tank
                tank = 0
        if total + tank < 0:
            return -1
        return start

File: leetcode/0134-gas-station/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 89 ms (83.45%)
     * MEMORY: 69.5 MB (55.78%)
    */
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int total{0}, tank{0}, start{0};
        for (int i=0; i < gas.size(); ++i){
            tank = tank + gas[i] - cost[i];
            if (tank < 0) {
                start = i + 1;
                total += tank;
                tank = 0;
            }
        }
        return total + tank < 0 ? -1 : start;
    }
};
